### ISSUE TITLE
Autograd precision for CONV + BN  between python version 1.11.0 and 1.12.1

### ISSUE THREAD

**Issue by Bestupidas** [pytorch/pytorch#87755]
### 🐛 Describe the bug

import torch

def compare_stat_dict_strict(stat_dict_1, stat_dict_2, rtol=5e-2, atol=1e-5):
    for k,v in stat_dict_1.items():
        if k not in stat_dict_2:
            raise KeyError
        else:
            np.testing.assert_allclose(to_numpy(v), to_numpy(stat_dict_2[k]), rtol=rtol, atol=atol)
if __name__ == "__main__":
  device='cpu'
  test_cycle = 2
  check_model=False
  np.random.seed(0)
  np_input = np.random.random((32, 3, 32, 32)).astype(np.float32)
  input_data = torch.from_numpy(np_input).to(device)
  
   # test model
  model = torch.nn.Sequential(
      torch.nn.Conv2d(3, 32, 3, 1, bias=False),
      torch.nn.BatchNorm2d(32)
  )
  model.train()
  model.to(device)
  loss_fun = torch.nn.MSELoss()
  

  save_model = False
  for _ in range(test_cycle):
      if save_model == False and not check_model:
          save_model = True
  
      if save_model:
          torch.save(model.state_dict(), "model_paramter.pth")
      else:
          load_dict = torch.load("model_paramter.pth", map_location=device)
          model.load_state_dict(load_dict, strict=True)
  
      model_output = model(input_data)
  
      if save_model:
          np.savez("model_output.npz", out=to_numpy(model_output))
      else:
          test_out = np.load('model_output.npz')['out']
          np.testing.assert_allclose(test_out, to_numpy(model_output))
  
      loss = loss_fun(torch.ones_like(model_output), model_output)
      print("loss=", loss)
      #empty grad
      for name, para in model.named_parameters():
          para.grad = None
      loss.backward()
  
      grad = {}
      for name, para in model.named_parameters():
          grad[name] = para.grad
      if save_model:
          torch.save(grad, "model_grad.pth")
      else:
          test_grad = torch.load("model_grad.pth")
          compare_stat_dict_strict(test_grad, grad, rtol=1e-7, atol=0)

I build a simple model with only one CONV and BN. But when using loss.backward() to generate the grad of parameters in model, there shows a precision of grad between  pytorch 1.11.0 and pytorch 1.12.1. the relative error of precision is about 4% in my test. Does my test has someting wrong or is it normal to has a small precision of autograd bewteen  pytorch 1.11.0 and 1.12.1 ?  

### Versions

autograd precision between pytorch 1.11.0 and 1.12.1 for CONV + BN

**Comment by soulitzer** [comment]
I'm not sure we promise bit-for-bit replicability between versions, but 4% doesn't seem normal.  Could you run:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

**Comment by Bestupidas** [comment]
Collecting environment information...
PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.24.2
Libc version: glibc-2.17

Python version: 3.7.13 (default, Mar 29 2022, 02:18:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-128-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 10.2.89
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: Graphics Device
Nvidia driver version: 440.33.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.12.1
[pip3] torchvision==0.13.1
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torchvision               0.13.1                   pypi_0    pypi

**Comment by Bestupidas** [comment]
import random
import numpy as np
import copy
import torch
import torch.nn as nn
import math
from torch.nn import functional as F
from ..base_network_blocks import BaseConv

class test_model(nn.Module):
    def __init__(self, layer_number):
        super(test_model, self).__init__()
        self.conv1 = nn.Sequential( nn.Conv2d(3,16,3,1, padding=1, bias=False),
                                    nn.BatchNorm2d(16),
                                    nn.ReLU())
        self.baseconv= nn.Sequential(nn.Conv2d(16,16,3,1, padding=1, bias=False),
                                     nn.BatchNorm2d(16),
                                     nn.ReLU())
        self.conv2 = nn.Sequential(
            *[self.baseconv for _ in range(layer_number)] )
    def forward(self, x):
        x = self.conv1(x)
        return self.conv2(x)

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)


def to_numpy(data):
    if isinstance(data, torch.Tensor):
        if data.requires_grad == True:
            return data.detach().cpu().numpy()
        else:
            return data.cpu().numpy()
    else:
        return np.array(data)

def save_model_grad_to_file(model, grad_file):
    grad = {}
    for name, para in model.named_parameters():
        grad[name] = para.grad
    torch.save(grad, grad_file)


def compare_stat_dict_strict(stat_dict_1, stat_dict_2, rtol=1e-7, atol=0):
    for k,v in stat_dict_1.items():
        if k not in stat_dict_2:
            raise KeyError
        else:
            np.testing.assert_allclose(to_numpy(v), to_numpy(stat_dict_2[k]), rtol=rtol, atol=atol)


def check_pytorch_autograd(input_data=None, model: torch.nn.Module= None, check_mode=False, device='cpu', epoch_number=2, model_layer=1):
    assert  epoch_number < 4
    #fix seed
    set_seed(0)
    #fix input
    if input_data is None:
        input_data = torch.rand(2,3,320,320, dtype=torch.float32)
    # test model
    if model == None:
        model = test_model(model_layer)
    model.to(device)
    model.train()
    loss_fun = torch.nn.L1Loss(reduction='mean')
    optim = torch.optim.SGD(model.parameters(), lr=0.1)
    test_flag = True
    for cur_epoch in range(epoch_number):
        model_paramter_path = "model_paramter_" + str(cur_epoch) + '.pth'
        if cur_epoch == 0:
            if not check_mode:
                torch.save(model.state_dict(), model_paramter_path)
            else:
                load_dict = torch.load(model_paramter_path, map_location=device)
                model.load_state_dict(load_dict, strict=True)

        model_output = model(input_data)

        model_output_path = "model_output_" + str(cur_epoch) + ".npz"
        if not check_mode:
            #model_output_path should be torch.Tensor, not list or tuple
            np.savez(model_output_path, out=to_numpy(model_output))
        else:
            test_out = np.load(model_output_path)['out']
            try:
                np.testing.assert_equal(test_out, to_numpy(model_output))
            except Exception as error:
                print(f"[cur_epoch={cur_epoch}, model output feature map check fail!!!]:{error}")
                test_flag = False
        loss = loss_fun(torch.ones_like(model_output), model_output)
        loss_path = "loss_" + str(cur_epoch) + ".npy"
        print(f"[cur_epoch={cur_epoch} loss={loss}]")
        if not check_mode:
            np.save(loss_path, to_numpy(loss))
        else:
            test_loss = np.load(loss_path)
            try:
                assert  np.isclose(to_numpy(loss), to_numpy(test_loss))
            except Exception as error:
                print(f"[cur_epoch={cur_epoch}, loss check fail!!!]:{error}")
                test_flag = False
        #empty grad
        optim.zero_grad()
        loss.backward()
        #compare grad
        grad = {}
        for name, para in model.named_parameters():
            grad[name] = para.grad
        model_grad_path = "model_grad_" + str(cur_epoch) + ".pth"
        if not check_mode:
            torch.save(grad, model_grad_path)
        else:
            test_grad = torch.load(model_grad_path)
            try:
                compare_stat_dict_strict(test_grad, grad, rtol=1e-7, atol=0)
            except Exception as error:
                print(f"[cur_epoch={cur_epoch}, model grad check fail!!!]:{error}")
                test_flag = False
        optim.step()
        #compare parameter
        optmizer_weight_path = "optimizer_weight_" + str(cur_epoch) + ".pth"
        if not check_mode:
            torch.save(model.state_dict(), optmizer_weight_path)
        else:
            optimizer_weight = torch.load(optmizer_weight_path)
            try:
                compare_stat_dict_strict(optimizer_weight, model.state_dict(), rtol=1e-7, atol=0)
            except Exception as error:
                print(f"[cur_epoch={cur_epoch}, optmizer update parameter check fail!!!]:{error}")
                test_flag = False

    if test_flag:
        print("test success !!!!")
    else:
        print("test fail !!!!")

    return test_flag

if __name__ == "__main__":
    check_pytorch_autograd(check_mode=False, model_layer=50)   #run this  to generate the reference data with pytorch 1.11.0
    check_pytorch_autograd(check_mode=True, model_layer=50)   #run this code with pytorch 1.21.1

    print("process done!!!")

**Comment by Bestupidas** [comment]
some modifications are made in my code to compare two epochs including autograd, parameter updating  of optimizer. the newest code is shown before. the model used is custom defined. the relative error is significant because the  absolute difference is a relative small value.  you can check the code and run it to has a test to get a similar result. hope for you reply 

[cur_epoch=0 loss=0.8217473030090332]
[cur_epoch=0, model grad check fail!!!]:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 427 / 432 (98.8%)
Max absolute difference: 6.7800283e-07
Max relative difference: 0.00107667
 x: array([[[[ 5.933375e-03, -4.279284e-02,  2.894593e-02],
         [ 1.683577e-02,  6.511120e-03,  6.215019e-02],
         [ 2.926406e-02, -1.439106e-02,  4.544495e-02]],...
 y: array([[[[ 5.933232e-03, -4.279268e-02,  2.894591e-02],
         [ 1.683561e-02,  6.510923e-03,  6.215017e-02],
         [ 2.926408e-02, -1.439120e-02,  4.544488e-02]],...
[cur_epoch=0, optmizer update parameter check fail!!!]:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 256 / 432 (59.3%)
Max absolute difference: 6.7055225e-08
Max relative difference: 2.481872e-05
 x: array([[[[ 0.001489,  0.111132, -0.159836],
         [-0.120909, -0.151404, -0.018465],
         [ 0.054336, -0.09068 , -0.158586]],...
 y: array([[[[ 0.001489,  0.111132, -0.159836],
         [-0.120909, -0.151404, -0.018465],
         [ 0.054336, -0.09068 , -0.158586]],...
[cur_epoch=1, model output feature map check fail!!!]:
Arrays are not equal

Mismatched elements: 1526112 / 3276800 (46.6%)
Max absolute difference: 0.00015065
Max relative difference: 1.
 x: array([[[[0.      , 0.      , 0.453038, ..., 0.      , 1.009565,
          0.037205],
         [0.203064, 0.464991, 0.343709, ..., 1.707108, 0.461237,...
 y: array([[[[0.      , 0.      , 0.453039, ..., 0.      , 1.009565,
          0.037202],
         [0.203067, 0.464987, 0.34371 , ..., 1.707106, 0.461236,...
[cur_epoch=1 loss=0.7675390839576721]
[cur_epoch=1, model grad check fail!!!]:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 432 / 432 (100%)
Max absolute difference: 1.9909348e-05
Max relative difference: 0.42751008
 x: array([[[[-2.711730e-03,  6.895724e-04,  6.398818e-04],
         [ 7.179931e-04,  1.159218e-03,  8.046153e-05],
         [-1.819707e-03,  2.009179e-03, -7.344221e-04]],...
 y: array([[[[-2.709164e-03,  6.913639e-04,  6.400633e-04],
         [ 7.155485e-04,  1.163871e-03,  8.199128e-05],
         [-1.827803e-03,  1.998825e-03, -7.280697e-04]],...
[cur_epoch=1, optmizer update parameter check fail!!!]:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 426 / 432 (98.6%)
Max absolute difference: 2.000481e-06
Max relative difference: 0.00087238
 x: array([[[[ 0.00176 ,  0.111063, -0.1599  ],
         [-0.120981, -0.15152 , -0.018473],
         [ 0.054518, -0.090881, -0.158512]],...
 y: array([[[[ 0.00176 ,  0.111063, -0.1599  ],
         [-0.12098 , -0.151521, -0.018473],
         [ 0.054518, -0.09088 , -0.158513]],...
test fail !!!!
process done!!!
